arch/platform/cooja/sys/cooja_mt.c

```diff
@@ -36,36 +36,95 @@
  * Fredrik Osterlind <fros@sics.se>
  */
 
-#include "contiki.h"
+#ifdef __APPLE__
+#define _DARWIN_C_SOURCE /* For pthread_get_stackaddr_np() */
+#else
+#define _GNU_SOURCE /* For pthread_getattr_np(). */
+#endif
+#include <pthread.h>
+
 #include "sys/cooja_mt.h"
-#include "sys/cc.h"
+
+#include <signal.h>
+#include <stdio.h>
 
 #define MT_STATE_READY   1
 #define MT_STATE_RUNNING 2
 #define MT_STATE_EXITED  5
 
 static struct cooja_mt_thread *current;
+static struct cooja_mt_thread *cooja_running_thread;
 
+#ifdef __APPLE__
+/* Avoid deprecated warnings for ucontext.h functions on OS X. */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+#endif
+
+/*--------------------------------------------------------------------------*/
+int
+cooja_mt_init(struct cooja_mt_thread *t)
+{
+#ifdef __APPLE__
+  /* Apple's makecontext() fails if size is less than MINSIGSTKSZ. */
+  if(COOJA_MTARCH_STACKSIZE < MINSIGSTKSZ) {
+    return -1;
+  }
+#endif
+  void *stackAddr;
+  size_t stackSize;
+  pthread_t self = pthread_self();
+#ifdef __APPLE__
+  stackAddr = pthread_get_stackaddr_np(self);
+  stackSize = pthread_get_stacksize_np(self);
+#else
+  pthread_attr_t attrs;
+  if(pthread_getattr_np(self, &attrs) != 0) {
+    return -2;
+  }
+  if(pthread_attr_getstack(&attrs, &stackAddr, &stackSize) != 0) {
+    return -3;
+  }
+#endif
+  getcontext(&t->ctxt);
+  t->ctxt.uc_stack.ss_sp = stackAddr;
+  t->ctxt.uc_stack.ss_flags = 0;
+  t->ctxt.uc_stack.ss_size = stackSize;
+  return 0;
+}
 /*--------------------------------------------------------------------------*/
 void
-cooja_mt_start(struct cooja_mt_thread *thread, void (* function)(void *), void *data)
+cooja_mt_start(struct cooja_mt_thread *caller,
+               struct cooja_mt_thread *t, void (*function)(void))
 {
-  /* Call the architecture dependant function to set up the processor
-     stack with the correct parameters. */
-  cooja_mtarch_start(&thread->thread, function, data);
+  for(int i = 0; i < COOJA_MTARCH_STACKSIZE; ++i) {
+    t->stack[i] = i;
+  }
 
-  thread->state = MT_STATE_READY;
+  if(getcontext(&t->ctxt) == -1) {
+    perror("getcontext failed");
+  }
```

Comment:
Should we not stop executing if `getcontext()` fails?

Reply:
There isn't any error checking/propagation in the old code, but I've added a return.